src/MC/fix_neighbor_swap.cpp

```diff
@@ -0,0 +1,949 @@
+               "Must specify at least 2 atom types in fix neighbor/swap 'types' keyword");
+}
+
+/* ---------------------------------------------------------------------- */
+
+int FixNeighborSwap::setmask()
+{
+  int mask = 0;
+  mask |= PRE_EXCHANGE;
+  return mask;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixNeighborSwap::init()
+{
+  c_pe = modify->get_compute_by_id("thermo_pe");
+  if (!c_pe) error->all(FLERR, Error::NOLASTLINE, "Could not find 'thermo_pe' compute");
+
+  c_voro = modify->get_compute_by_id(id_voro);
+  if (!c_voro)
+    error->all(FLERR, Error::NOLASTLINE, "Could not find voronoi compute ID {}", id_voro);
+
+  if (nswaptypes < 2 && !diff_flag)
+    error->all(FLERR, "Must specify at least 2 types in fix neighbor/swap command");
+
+  // set index and check validity of region
+
+  if (idregion) {
+    region = domain->get_region_by_id(idregion);
+    if (!region)
+      error->all(FLERR, Error::NOLASTLINE, "Region {} for fix neighbor/swap does not exist",
+                 idregion);
+  }
+
+  for (int iswaptype = 0; iswaptype < nswaptypes; iswaptype++)
+    if (type_list[iswaptype] <= 0 || type_list[iswaptype] > atom->ntypes)
+      error->all(FLERR, "Invalid atom type in fix neighbor/swap command");
+
+  int *type = atom->type;
+  if (atom->q_flag) {
+    double qmax, qmin;
+    int firstall, first;
+    memory->create(qtype, nswaptypes, "neighbor/swap:qtype");
+    for (int iswaptype = 0; iswaptype < nswaptypes; iswaptype++) {
+      first = 1;
+      for (int i = 0; i < atom->nlocal; i++) {
+        if (atom->mask[i] & groupbit) {
+          if (type[i] == type_list[iswaptype]) {
+            if (first > 0) {
+              qtype[iswaptype] = atom->q[i];
+              first = 0;
+            } else if (qtype[iswaptype] != atom->q[i])
+              first = -1;
+          }
+        }
+      }
+      MPI_Allreduce(&first, &firstall, 1, MPI_INT, MPI_MIN, world);
+      if (firstall < 0)
+        error->all(FLERR, Error::NOLASTLINE,
+                   "All atoms of a swapped type must have the same charge");
+      if (firstall > 0)
+        error->all(FLERR, Error::NOLASTLINE,
+                   "At least one atom of each swapped type must be present to define charges");
+      if (first) qtype[iswaptype] = -DBL_MAX;
+      MPI_Allreduce(&qtype[iswaptype], &qmax, 1, MPI_DOUBLE, MPI_MAX, world);
+      if (first) qtype[iswaptype] = DBL_MAX;
+      MPI_Allreduce(&qtype[iswaptype], &qmin, 1, MPI_DOUBLE, MPI_MIN, world);
+      if (qmax != qmin)
+        error->all(FLERR, Error::NOLASTLINE, "All atoms of a swapped type must have same charge.");
+    }
+  }
+
+  memory->create(sqrt_mass_ratio, atom->ntypes + 1, atom->ntypes + 1,
+                 "neighbor/swap:sqrt_mass_ratio");
+  for (int itype = 1; itype <= atom->ntypes; itype++)
+    for (int jtype = 1; jtype <= atom->ntypes; jtype++)
+      sqrt_mass_ratio[itype][jtype] = sqrt(atom->mass[itype] / atom->mass[jtype]);
```

Comment:
There should be an error if atoms have an `rmass` instead of a type mass.

Reply:
@jtclemm good catch! thanks. I just added code that makes this fix handle per-atom-masses, too.